### PR TITLE
Adapt pathfinding to use new infra

### LINKED ISSUE #721: Signalization cleanup: adapt pathfinding to new infra, remove old infra

### PR DESCRIPTION
**Pull request by eckter**
Fixes #721

### CODE REVIEW

File: core/src/main/java/fr/sncf/osrd/railjson/schema/schedule/RJSTrainSchedule.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+    public ID<RJSRoute>[] routes;
+
+    /** List of stops */
+    public RJSTrainStop[] stops;
+
+    /** Identifier of the previous train, if any.
+     * If specified, this train will only leave after the previous train reached its destination */
+    @Json(name = "previous_train_id")
+    public String previousTrainId;
```
Comment (multun) [inline core/src/main/java/fr/sncf/osrd/railjson/schema/schedule/RJSTrainSchedule.java]: I'm confused, is it used yet?

File: core/src/main/java/fr/sncf/osrd/utils/new_graph/Pathfinding.java [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+
+@SuppressFBWarnings(
+        value = "FE_FLOATING_POINT_EQUALITY",
+        justification = "No arithmetic is done on values where we test for equality, only copies"
+)
+public class Pathfinding<NodeT, EdgeT> {
+
+    /** Pathfinding step */
+    private record Step<EdgeT>(
+            EdgeRange<EdgeT> range, // Next step
```
Comment (multun) [inline core/src/main/java/fr/sncf/osrd/utils/new_graph/Pathfinding.java]: the comment there seems off

File: core/src/main/java/fr/sncf/osrd/utils/new_graph/Pathfinding.java [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+            Step<EdgeT> prev, // Previous step (to construct the result)
+            double totalDistance, // Total distance from the start
+            double weight, // Priority queue weight (could be different from totalDistance to allow for A*)
```
Comment (multun) [inline core/src/main/java/fr/sncf/osrd/utils/new_graph/Pathfinding.java]: can't you just remove the totalDistance field?

File: core/src/main/java/fr/sncf/osrd/utils/new_graph/Pathfinding.java [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+            var startRange = new EdgeRange<>(location.edge, location.offset, edgeToLength.apply(location.edge));
+            registerStep(startRange, null, 0, 0);
+        }
+        while (true) {
+            var step = queue.poll();
+            if (step == null)
+                return null;
+            if (hasReachedEnd(locations, step))
+                return buildResult(step);
+            for (var stopLocation : locations.get(step.nReachedSteps + 1))
```
Comment (multun) [inline core/src/main/java/fr/sncf/osrd/utils/new_graph/Pathfinding.java]: it'd be nice to have a little comment above this line, like: "check if this step reaches the current target"

File: core/src/main/java/fr/sncf/osrd/utils/new_graph/Pathfinding.java [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+)
+public class Pathfinding<NodeT, EdgeT> {
+
+    /** Pathfinding step */
+    private record Step<EdgeT>(
+            EdgeRange<EdgeT> range, // Next step
+            Step<EdgeT> prev, // Previous step (to construct the result)
+            double totalDistance, // Total distance from the start
+            double weight, // Priority queue weight (could be different from totalDistance to allow for A*)
+            int nReachedSteps // How many steps have been passed
```
Comment (multun) [inline core/src/main/java/fr/sncf/osrd/utils/new_graph/Pathfinding.java]: in this file, step means either a pathfinding step, or a goal. that's confusing, and I'd very much like that we use a convention which does not use a single words for two distinct things. how about calling the targets (the things not in the queue) goals or just target?

File: core/src/main/java/fr/sncf/osrd/utils/new_graph/Pathfinding.java [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+    private boolean hasReachedEnd(
+            List<Collection<EdgeLocation<EdgeT>>> locations,
+            Step<EdgeT> step
+    ) {
+        return step.nReachedSteps >= locations.size() - 1;
+    }
+
+    /** Builds the result, iterating over the previous steps and merging ranges */
+    private List<EdgeRange<EdgeT>> buildResult(Step<EdgeT> step) {
+        var invertedResult = new ArrayList<EdgeRange<EdgeT>>();
```
Comment (multun) [inline core/src/main/java/fr/sncf/osrd/utils/new_graph/Pathfinding.java]: you can use an arraydeque and add to the front to make iteration slightly easier (it has the same complexity / almost same runtime code)

**Review by multun — CHANGES_REQUESTED** [review]

File: core/src/main/java/fr/sncf/osrd/railjson/schema/schedule/RJSTrainSchedule.java (hunk 1 above)
Comment (eckter) [reply core/src/main/java/fr/sncf/osrd/railjson/schema/schedule/RJSTrainSchedule.java]: No, i just restored the file as it was before (minus what doesn't compile). I use it for some tests. 

I could remove the unused fields, but I don't really see the point. It still describes the format used by the generator.

File: core/src/main/java/fr/sncf/osrd/utils/new_graph/Pathfinding.java (hunk 3 above)
Comment (eckter) [reply core/src/main/java/fr/sncf/osrd/utils/new_graph/Pathfinding.java]: I plan on switching to an A* relatively soon where this distinction is needed. Having this in mind from the start makes it easier to write

Comment (multun) [reply core/src/main/java/fr/sncf/osrd/utils/new_graph/Pathfinding.java]: as far as I know A* only needs a weight, not the total distance

Comment (eckter) [reply core/src/main/java/fr/sncf/osrd/utils/new_graph/Pathfinding.java]: In A*, weigh = distance + heuristic, but you need the distance alone to compute the distance of the next step. I use `step.totalDistance` at several places in the code and those would be wrong if it included the heuristic

File: core/src/main/java/fr/sncf/osrd/railjson/schema/schedule/RJSTrainSchedule.java (hunk 1 above)
Comment (multun) [reply core/src/main/java/fr/sncf/osrd/railjson/schema/schedule/RJSTrainSchedule.java]: I don't recall having ever used this field, I really think it should go if we don't use it. someone might read the file and assume we do something with this data

**Review by multun — APPROVED** [review]

**Pull request merged**